### PR TITLE
plugin/pkg/auth/authenticator/token/oidc: get groups from custom claim

### PR DESCRIPTION
**Pull request by ericchiang**
This PR provides functionality that allows the current OpenID Connect plugin to optionally read user group membership from a custom JWT claim.

OpenID Connect permits custom claims as part of ID Tokens ([document here](https://openid.net/specs/openid-connect-basic-1_0.html#IDToken)). And this allows Kubernetes aware OpenID Connect providers to produce ID Tokens with group information.

cc: @erictune, @philips, @bobbyrullo

### CODE REVIEW

**Comment by brendandburns** [comment]
reassigning to @erictune because auth.

@liggitt may also be interested.

File: plugin/pkg/auth/authenticator/token/oidc/oidc.go [hunk 1]
```diff
@@ -155,8 +156,19 @@ func (a *OIDCAuthenticator) AuthenticateToken(value string) (user.Info, bool, er
 		username = fmt.Sprintf("%s#%s", a.clientConfig.ProviderConfig.Issuer, claim)
 	}
 
-	// TODO(yifan): Add UID and Group, also populate the issuer to upper layer.
-	return &user.DefaultInfo{Name: username}, true, nil
+	// TODO(yifan): Add UID, also populate the issuer to upper layer.
+	info := &user.DefaultInfo{Name: username}
+
+	if a.groupsClaim != "" {
+		groups, ok, err := claims.StringsClaim(a.groupsClaim)
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/oidc/oidc.go]: how does this handle a missing claim? I'd expect that to result in an empty groups list, not an auth error

Comment (ericchiang) [inline plugin/pkg/auth/authenticator/token/oidc/oidc.go]: The only error scenario is when the claim is not an array of strings.

```
{}                               // No groups claim, no error.
{"groups": ["group1", "group2"]} // Good response, no error.
{"groups": "group1"}             // Malformed groups claim, this causes an error.
```

Idea being that if authn sees a custom groups claim, it's particular about how that claim is formatted.

We could just log an error and continue though, either way.

Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/oidc/oidc.go]: ah, ok, didn't dig into StringsClaim

Comment (ericchiang) [inline plugin/pkg/auth/authenticator/token/oidc/oidc.go]: Yeah sorry it's not super obvious. I'll add a comment.

File: cmd/kube-apiserver/app/options/options.go [hunk 2]
```diff
@@ -165,6 +166,7 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.OIDCUsernameClaim, "oidc-username-claim", "sub", ""+
 		"The OpenID claim to use as the user name. Note that claims other than the default ('sub') is not "+
 		"guaranteed to be unique and immutable. This flag is experimental, please see the authentication documentation for further details.")
+	fs.StringVar(&s.OIDCGroupsClaim, "oidc-groups-claim", "", "If provided, the name of a custom OpenID Connect claim for specifying user groups. The claim value is expected to be an array of strings. This flag is experimental, please see the authentication documentation for further details.")
```
Comment (erictune) [inline cmd/kube-apiserver/app/options/options.go]: I take it that Dex is going to be the only IdP that implements this at first?

Comment (erictune) [inline cmd/kube-apiserver/app/options/options.go]: Oh, sounds like Azure supports groups, from this article:
http://www-01.ibm.com/support/knowledgecenter/SSEQTP_8.5.5/com.ibm.websphere.wlp.doc/ae/twlp_config_oidc_rp.html

Cool.

**Comment by erictune** [comment]
lgtm

**Comment by erictune** [comment]
add to whitelist

**Comment by erictune** [comment]
@k8s-merge-robot add-to-whitelist

File: cmd/kube-apiserver/app/options/options.go (hunk 2 above)
Comment (gambol99) [inline cmd/kube-apiserver/app/options/options.go]: [Keycloak](https://github.com/keycloak/keycloak) would support this via the "group membership" client mapper as well ... Definitely worth a mention :-)

Comment (therc) [inline cmd/kube-apiserver/app/options/options.go]: @erictune any chance you could lobby with Eric Sachs to have GApps support it? Right now you need to call the Directory API to obtain group membership information...

**Pull request merged**